pkg/controller/advanceddeployment/reconciler.go

```diff
@@ -0,0 +1,158 @@
+package advanceddeployment
+
+import (
+	"context"
+	"encoding/json"
+	"flag"
+	"reflect"
+	"time"
+
+	"github.com/openkruise/rollouts/api/v1alpha1"
+	"github.com/openkruise/rollouts/pkg/controller/advanceddeployment/util"
+	apps "k8s.io/api/apps/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/klog/v2"
+	"k8s.io/utils/pointer"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	DefaultDuration = time.Second
+)
+
+var (
+	concurrentReconciles = 3
+)
+
+func init() {
+	flag.IntVar(&concurrentReconciles, "advanced-deployment-workers", 3, "Max concurrent workers for advanced deployment controller.")
+}
+
+// Add creates a new Rollout Controller and adds it to the Manager with default RBAC. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	return add(mgr, newReconciler(mgr))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	recorder := mgr.GetEventRecorderFor("advanced-deployment-controller")
+	cli := mgr.GetClient()
+	return &AdvancedDeploymentReconciler{
+		Client:   cli,
+		Scheme:   mgr.GetScheme(),
+		recorder: recorder,
+	}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("advanced-deployment-controller", mgr, controller.Options{
+		Reconciler: r, MaxConcurrentReconciles: concurrentReconciles})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to Deployment
+	err = c.Watch(&source.Kind{Type: &apps.Deployment{}}, &handler.EnqueueRequestForObject{}, predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			oldObject := e.ObjectOld.(*apps.Deployment)
```

Comment:
How do you know which Deployment should perform the Advanced Deployment Reconcile?

Reply:
sounds reasonable